app/workers/import_ml_data_worker.rb

```diff
@@ -0,0 +1,69 @@
+# frozen_string_literal: true
+
+require 'csv'
+# Worker to bulk import Machine Learnt (ML) extracts from web accessibly located csv
+class ImportMLDataWorker
+  include Sidekiq::Worker
+  sidekiq_options retry: 2
+
+  def perform(csv_filepath, workflow_id)
+    workflow = Workflow.find(workflow_id)
+    upsert_extracts csv_filepath, workflow
+    run_workflow_reducers(workflow) if workflow&.reducers&.any?
+    project = Project.find(workflow.project_id)
+    run_project_reducers(project) if project&.has_reducers?
+  end
+
+  private
+
+  def upsert_extracts(csv_filepath, workflow_id)
+    UrlDownloader.stream(csv_filepath) do |io|
+      csv = CSV.new(io, headers: true)
+      csv.each do |row|
+        hashed_row = row.to_hash
+        upsert_subject hashed_row['subject_id']
+        extract = init_extract hashed_row, workflow_id
+        set_extract_data_from_row hashed_row, extract
+        extract.save!
+      end
+    end
+  end
+
+  def upsert_subject(subject_id)
+    s = Subject.where(id: subject_id).first_or_initialize
+    s.metadata = {} if s.metadata.nil?
+    s.save!
+  end
+
+  def init_extract(hashed_row, workflow)
+    Extract.where(
+      workflow_id: workflow.id,
+      subject_id: hashed_row['subject_id'],
+      extractor_key: hashed_row['extractor_key'],
+      workflow_version: hashed_row['workflow_version'],
```

Comment:
if any of these values are nil / missing in the hashed_row definition - will that then modify the lookup. E.g. `hashed_row['workflow_version'] = nil` then Extract.where(workflow_version: nil) is different to not set in the data. 

This might not be an issue but i can see the where clause returning different data due to a nil presence in the lookup hash vs not present.

Perhaps create the list of lookup params and then compact them to ensure you only lookup records with the values that are present. E.g. 
```
# construct the params
extract_find_params = { .... }.compact
Extract.where(extract_find_params).first_or_initialize
```